src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,315 @@
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      options?.expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    const updatedStream = await this.collection.findOneAndUpdate(
+      {
+        streamName: { $eq: streamName },
+        events: { $size: stream.events.length },
+        updatedAt: new Date(),
+      },
+      { $push: { events: { $each: eventCreateInputs } } },
+      { returnDocument: 'after' },
+    );
+
+    if (!updatedStream) {
+      const currentStream = await this.collection.findOne({
+        streamName: { $eq: streamName },
+      });
+      throw new ExpectedVersionConflictError(
+        currentStream?.events.length ?? -1,
+        stream.events.length,
+      );
+    }
+
+    const { streamType, entityId } = fromStreamName(streamName);
+
+    if (options?.projections) {
+      await Promise.all(
+        options.projections.map((project) =>
+          project({
+            streamName,
+            streamType,
+            entityId,
+            streamVersion: updatedStream.events.length,
+            events: updatedStream.events.map(this.parseEvent(streamName)),
+          }),
+        ),
+      );
+    }
+
+    if (options?.asyncProjections) {
+      for (const project of options.asyncProjections) {
+        project({
+          streamName,
+          streamType,
+          entityId,
+          streamVersion: updatedStream.events.length,
+          events: updatedStream.events.map(this.parseEvent(streamName)),
+        });
+      }
+    }
+
+    return {
+      nextExpectedStreamVersion: updatedStream.events.length,
+      createdNewStream,
+    };
+  }
+
+  /**
+   * Transforms the `event` from the saved format into the usable object
+   * at runtime. This function may be altered later to match `stringifyEvent`.
+   */
+  private parseEvent<EventType extends Event>(streamName: StreamName) {
+    return (
+      event: EventStreamEvent,
+      index?: number,
+    ): ReadEvent<EventType, ReadEventMetadata> => {
+      const metadata = {
+        ...JSON.parse(event.metadata),
+        eventId: event._id,
+        streamName,
+        streamPosition: BigInt(index ?? 0),
+      } satisfies ReadEventMetadata;
+
+      // TODO: is this correct?
+      // @ts-expect-error
+      return {
+        __brand: 'Event',
+        type: event.type,
+        data: JSON.parse(event.data),
+        metadata,
+      };
+    };
+  }
+
+  /**
+   * Transforms the `event` into a saveable format. This function may
+   * be altered later depending on storage needs.
+   */
+  private stringifyEvent<EventType extends Event>(
+    event: EventType,
+  ): EventStreamEvent {
+    return {
+      _id: new ObjectId(),
+      type: event.type,
+      data: JSON.stringify(event.data),
+      metadata: JSON.stringify(
+        event.metadata ?? {
+          now: new Date(),
+        },
+      ),
+    };
+  }
+}
+
+export const getMongoDBEventStore = async (
```

Comment:
I know the other event stores are sync here, but I want to add the `streamName` unique index which is an async call.